apisix/plugins/recaptcha.lua

```diff
@@ -0,0 +1,122 @@
+local radix = require("resty.radixtree")
+local core = require("apisix.core")
+local http = require("resty.http")
+
+local schema = {
+    type = "object",
+    properties = {
+        recaptcha_secret_key = { type = "string" },
+        apis = {
+            type = "array",
+            items = {
+                type = "object",
+                properties = {
+                    path = { type = "string" },
+                    methods = { type = "array", items = { type = "string" }, minItems = 1 },
+                    param_from = { type = "string", default = "header", enum = { "header", "query" } },
+                    param_name = { type = "string", default = "captcha" },
+                }
+            },
+            minItems = 1
+        },
+        response = {
+            type = "object",
+            properties = {
+                content_type = { type = "string", default = "application/json; charset=utf-8" },
+                status_code = { type = "number", default = 400 },
+                body = { type = "string", default = '{"message": "invalid captcha"}' }
+            }
+        },
+
+    },
+    additionalProperties = false,
+    required = { "recaptcha_secret_key" },
+}
+
+local recaptcha_url = "https://www.recaptcha.net"
+
+local _M = {
+    version = 0.1,
+    priority = 700,
+    name = "recaptcha",
+    schema = schema,
+}
+
+function _M.check_schema(conf, schema_type)
+    return core.schema.check(schema, conf)
+end
+
+local function build_radixtree(apis)
+    local items = {}
+    for _, api in ipairs(apis) do
+        local item = {
+            paths = { api.path },
+            methods = api.methods,
+            metadata = api,
+        }
+        table.insert(items, item)
+    end
+    return radix.new(items)
+end
+
+local function find_api(request, apis)
+    local rx = build_radixtree(apis)
+    return rx:match(request.path, { method = request.method })
+end
```

Comment:
Why would you want to manage the api manually through your plugin? With all due respect, this does not match the design of APISIX.

`Route` should be set up and plugins enabled for them first, rather than having your global plugins manage the routes.

Reply:
Good question. 
This is designed on purpose. In enterprise architecture, you may have hundreds of microservice in the backend. the api gateway dispatches the requests to those services based on the prefix of URL.

For example
```
/user/* -> user service
/book/* -> book service
...
```
Let's say if every service has one or two API need to be verified by recaptcha. If we try to add recaptcha plugin in every service/route, I think it will end up a mess, and it's not good for maintaining the plugin(or secret key in the plugin) in one place. 

perhaps we can add an option property to verify every request once the plugin is executed. just like the other plugins.

Reply:
I think this is creating a model that is rarely used and doesn't fit the existing concept and needs @spacewander to be looked at.

Reply:
I prefer to add an expression from lua-resty-expr to filter out the requests that need to be reCAPTCHA, instead of using a mini router matcher (which is an anti-pattern).

Reply:
Actually, it isn't a problem as we can configure the basic plugins in service and configure reCAPTCHA in route.

Reply:
@spacewander I just dig into lua-resty-expr document, and found that it seems only support arg_xxx now(please correct me if I'm wrong). Usually, the recaptcha API depends on the path rather than arg.

And yes, this is an anti-pattern. I have discussed this with @shuaijinchao before. To match APISIX pattern, I can add a match_all(boolean, default true) property to verify every request once the plugin is executed. Just like the other plugins.

What do you think?

Reply:
> Actually, it isn't a problem as we can configure the basic plugins in service and configure reCAPTCHA in route.

Sorry, any example of this?